In [3]:
import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# from sklearn import ?
# from sklearn.metrics import ?

<br>

## 1. Preparing dataset (2번부터 실습 진행)

In [21]:
titan_df = pd.read_csv('titanic.csv')
titan_df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [22]:
titan_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [23]:
titan_df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

- train data의 표본 샘플은 타이나닉호에 탑승한 실제 승객 수(2,224명)의 40%인 891명이다. 
- 표본 샘플의 평균 생존율은 약 38%이다. 
- 대부분의 승객(75% 이상)은 부모나 자녀와 함께 탑승하지 않았다. 
- 지불 요금은 상당히 다양하게 나타나며, 512달러 만큼 높은 요금을 지불하는 경우는 거의 없었다.(1% 미만) 
- 65-80세 범위에 속하는 노인은 거의 없다. (1% 미만)

## Data info

- **PassengerId** : Unique ID of passenger
- **Survived** : 0 = No, 1 = Yes
- **pclass** : Ticket class (1 = 1st, 2 = 2nd, 3 = 3rd)
- **sibsp** : # of siblings & spouses aboard the Titanic
- **parch** : # of parents / children aboard the Titanic
- **ticket** : Ticket number
- **cabin** : Cabin number
- **embarked** : Port of Embarkation (C = Cherbourg, Q = Queenstown, S = Southampton)

In [5]:
y_data = data_df[['Survived']]
# y_data.head(3)

In [6]:
del data_df['Survived']
x_data = data_df.copy()
# x_data.head(3)

<br>

## 2. Feature engineering & Feature selection

#### 시도해볼 수 있는 전략들

- 불필요한 열이나 예측에 방해가 되는 열은 아예 지우기 (ex. PassengerId)
- 결측치 채우기 
- Text로 되어있는 Category(Factor)는 숫자로 바꿔주기 (ex. Male/Female -> 0/1)
- 실수 범위를 구간 범위로 바꿔주기 
- 필요한 경우 기존 열을 바탕으로 새로운 열을 계산해 추가하기

ex. 보험사에서 다루는 고객데이터셋 중 나이열에 있는 결측치를 다루려 할 때
- 랜덤 값으로 채우기
- 대표 값으로 채우기 (mean평균, median중위값, mode최빈값)
- 위아래 데이터의 평균값으로 채우기
- 행들을 그룹으로 묶은 후 해당 그룹의 대표값으로 채우기
- 머신러닝 예측모델을 만들어 해당 열의 결측치를 채우기
- 특정 기준 이상의 결측치가 있을 시 해당 열 삭제

In [19]:
x_data.head(50)

PassengerId  Pclass                                               Name  \
0             1       3                            Braund, Mr. Owen Harris   
1             2       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2             3       3                             Heikkinen, Miss. Laina   
3             4       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4             5       3                           Allen, Mr. William Henry   
5             6       3                                   Moran, Mr. James   
6             7       1                            McCarthy, Mr. Timothy J   
7             8       3                     Palsson, Master. Gosta Leonard   
8             9       3  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)   
9            10       2                Nasser, Mrs. Nicholas (Adele Achem)   
10           11       3                    Sandstrom, Miss. Marguerite Rut   
11           12       1                           Bonnell, Miss. Elizabeth   
12           13       3                     Saundercock, Mr. William Henry   
13           14       3                        Andersson, Mr. Anders Johan   
14           15       3               Vestrom, Miss. Hulda Amanda Adolfina   
15           16       2                   Hewlett, Mrs. (Mary D Kingcome)    
16           17       3                               Rice, Master. Eugene   
17           18       2                       Williams, Mr. Charles Eugene   
18           19       3  Vander Planke, Mrs. Julius (Emelia Maria Vande...   
19           20       3                            Masselmani, Mrs. Fatima   
20           21       2                               Fynney, Mr. Joseph J   
21           22       2                              Beesley, Mr. Lawrence   
22           23       3                        McGowan, Miss. Anna "Annie"   
23           24       1                       Sloper, Mr. William Thompson   
24           25       3                      Palsson, Miss. Torborg Danira   
25           26       3  Asplund, Mrs. Carl Oscar (Selma Augusta Emilia...   
26           27       3                            Emir, Mr. Farred Chehab   
27           28       1                     Fortune, Mr. Charles Alexander   
28           29       3                      O'Dwyer, Miss. Ellen "Nellie"   
29           30       3                                Todoroff, Mr. Lalio   
30           31       1                           Uruchurtu, Don. Manuel E   
31           32       1     Spencer, Mrs. William Augustus (Marie Eugenie)   
32           33       3                           Glynn, Miss. Mary Agatha   
33           34       2                              Wheadon, Mr. Edward H   
34           35       1                            Meyer, Mr. Edgar Joseph   
35           36       1                     Holverson, Mr. Alexander Oskar   
36           37       3                                   Mamee, Mr. Hanna   
37           38       3                           Cann, Mr. Ernest Charles   
38           39       3                 Vander Planke, Miss. Augusta Maria   
39           40       3                        Nicola-Yarred, Miss. Jamila   
40           41       3     Ahlin, Mrs. Johan (Johanna Persdotter Larsson)   
41           42       2  Turpin, Mrs. William John Robert (Dorothy Ann ...   
42           43       3                                Kraeff, Mr. Theodor   
43           44       2           Laroche, Miss. Simonne Marie Anne Andree   
44           45       3                      Devaney, Miss. Margaret Delia   
45           46       3                           Rogers, Mr. William John   
46           47       3                                  Lennon, Mr. Denis   
47           48       3                          O'Driscoll, Miss. Bridget   
48           49       3                                Samaan, Mr. Youssef   
49           50       3      Arnold-Franchi, Mrs. Josef (Josefine Franchi)   

       Sex   Age  SibSp  

In [7]:
x_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Pclass       891 non-null    int64  
 2   Name         891 non-null    object 
 3   Sex          891 non-null    object 
 4   Age          714 non-null    float64
 5   SibSp        891 non-null    int64  
 6   Parch        891 non-null    int64  
 7   Ticket       891 non-null    object 
 8   Fare         891 non-null    float64
 9   Cabin        204 non-null    object 
 10  Embarked     889 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 76.7+ KB


In [9]:
x_data.isna().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [18]:
pd.set_option('display.max_rows', None)
x_data['Cabin'].value_counts()

B96 B98            4
G6                 4
C23 C25 C27        4
C22 C26            3
F33                3
F2                 3
E101               3
D                  3
C78                2
C93                2
E8                 2
D36                2
B77                2
C123               2
E121               2
E44                2
D35                2
C125               2
E67                2
B35                2
B18                2
E24                2
B49                2
C65                2
B20                2
B5                 2
B57 B59 B63 B66    2
C126               2
B51 B53 B55        2
F4                 2
C124               2
F G73              2
B58 B60            2
C52                2
D33                2
C68                2
D20                2
D26                2
B28                2
C83                2
E25                2
D17                2
B22                2
C92                2
C2                 2
E33                2
C70                1
E58          

In [17]:
x_data['Embarked'].value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

In [10]:
# # 기본적인 결측치 채우는 방법
# mean_age = x_data['Age'].median(skipna=True)
# x_data['Age'] = x_data['Age'].fillna(mean_age) # or .fillna(int/float/str 값)

<br>

## 2. Train - Test split (비율을 7:3 으로 유지해주시고, seed는 0을 적용해주세요)

<br>

## 3. Create model instance variable (동시에 여러 모델을 다른 이름으로 만들 수 있습니다.)

<br>

## 4. Train the model

<br>

## 5. Predict on test data & Check the result with metrics (모델 간 비교가 가능합니다.)